src/librustc/traits/error_reporting.rs

```diff
@@ -684,24 +684,36 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                                    expected_found.found,
                                                    expected_trait_ty.is_closure())
                 } else if let &TypeError::Sorts(ref expected_found) = e {
-                    let expected = if let ty::TyTuple(tys, _) = expected_found.expected.sty {
+                    let expected_len = if let ty::TyTuple(tys, _) = expected_found.expected.sty {
```

Comment:
Huh. I am a bit dubious about this `if` in the first-place. I guess I have to run the code -- what happens if you just remove this logic altogether, though? It feels very odd to me to have to do all this matching.